smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/traits/StandardRegionalEndpointsTrait.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Override
+    protected Node createNode() {
+        ObjectNode.Builder partitionSpecialCasesNodeBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<String, List<PartitionSpecialCase>> entry: partitionSpecialCases.entrySet()) {
+            List<Node> nodes = new ArrayList<>();
+            for (PartitionSpecialCase partitionSpecialCase : entry.getValue()) {
+                nodes.add(partitionSpecialCase.toNode());
+            }
+            partitionSpecialCasesNodeBuilder.withMember(entry.getKey(), Node.fromNodes(nodes));
+        }
+
+        ObjectNode.Builder regionSpecialCasesNodeBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<String, List<RegionSpecialCase>> entry: regionSpecialCases.entrySet()) {
+            List<Node> nodes = new ArrayList<>();
+            for (RegionSpecialCase regionSpecialCase : entry.getValue()) {
+                nodes.add(regionSpecialCase.toNode());
+            }
+            regionSpecialCasesNodeBuilder.withMember(entry.getKey(), Node.fromNodes(nodes));
+        }
+
+        return ObjectNode.objectNodeBuilder()
+                .sourceLocation(getSourceLocation())
+                .withMember(PARTITION_SPECIAL_CASES, partitionSpecialCasesNodeBuilder.build())
+                .withMember(REGION_SPECIAL_CASES, partitionSpecialCasesNodeBuilder.build())
+                .build();
+    }
+
+    @Override
+    public SmithyBuilder<StandardRegionalEndpointsTrait> toBuilder() {
+        return new Builder()
+                .partitionSpecialCases(partitionSpecialCases)
+                .regionSpecialCases(regionSpecialCases);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode objectNode = value.expectObjectNode();
+
+            StandardRegionalEndpointsTrait.Builder builder = builder()
+                    .sourceLocation(value);
+
+            Optional<ObjectNode> memberNode = objectNode.getObjectMember(PARTITION_SPECIAL_CASES);
+            if (memberNode.isPresent()) {
+                for (Map.Entry<String, Node> entry
+                        : objectNode.expectObjectMember(PARTITION_SPECIAL_CASES).getStringMap().entrySet()) {
+                    List<PartitionSpecialCase> partitionSpecialCases = new ArrayList<>();
+                    for (Node node: entry.getValue().expectArrayNode().getElements()) {
+                        partitionSpecialCases.add(PartitionSpecialCase.fromNode(node));
+                    }
+                    builder.putPartitionSpecialCases(
+                        entry.getKey(), Collections.unmodifiableList(partitionSpecialCases));
+                }
+            }
+
+            memberNode = objectNode.getObjectMember(REGION_SPECIAL_CASES);
```

Comment:
Should use `objectNode.containsMember` in the if statement, since this `Optional` value is discarded.